docs(css): FF126 - Support for `shape()` function

dipikabh opened the pull request:
## 🎯 Target release date

📆 May 14, 2024

## Description

Firefox 126 adds support for a new function in the `<basic-shape>` data type: `shape()`.

The support is currently available behind the preference `layout.css.basic-shape-shape.enabled`, which is enabled by default in Nightly.

**Specification**: https://drafts.csswg.org/css-shapes-2/#shape-function

**Firefox bugs:**
- [[css-shapes-2] Implement shape() function for clip-path (bug 1823463)](https://bugzilla.mozilla.org/show_bug.cgi?id=1823463)
- [[css-shapes-2] Implement shape() function for offset-path (bug 1884424)](https://bugzilla.mozilla.org/show_bug.cgi?id=1884424)
- [[css-shapes-2] Support interpolation between shape() and path() (bug 1884425)](https://bugzilla.mozilla.org/show_bug.cgi?id=1884425)

## Notes for reviewers

1. You can start your review with this new page: `/en-us/Web/CSS/basic-shape/shape`
   - The "Specifications" and "Browser compatibility" sections are yet to be updated, so they will show as broken at the moment.
2. Next page that can be reviewed: `/en-us/Web/CSS/basic-shape`
   - Added the new bit about `shape()`
   - Updated the [Interpolation](https://developer.mozilla.org/en-US/docs/Web/CSS/basic-shape#interpolation_of_basic_shapes) section to add info about `shape()`
   (section in spec: https://drafts.csswg.org/css-shapes-2/#interpolating-shape)
3. `/en-US/docs/Web/CSS/clip-path` and `/en-US/docs/Web/CSS/offset-path` have been updated to add a reference to the `shape()` function.
4. `/en-us/Web/CSS/basic-shape/circle`: Added info about the circle function
5. `/en-us/Web/CSS/path`: Minor update in syntax formatting
6. `/en-us/Web/SVG/Attribute/d`: Added anchor links to the command sections to aid in navigation and updated few headings for consistency

## Additional details

Coming soon:
- PR to update BCD
- PR to add Firefox release note

## Related issues and pull requests

Doc issue tracker: https://github.com/mdn/content/issues/33081

dipikabh committed:
wip

dipikabh committed:
tests link updates

dipikabh committed:
adds more inline links)

dipikabh committed:
test links in shape page

dipikabh committed:
adds reminaing links to path commands

dipikabh committed:
updates other basic-shape pages

dipikabh committed:
undo's the alphabetical ordering of shape functions

dipikabh committed:
updates fill-rule example syntax

dipikabh committed:
fixes shape() link

dipikabh committed:
fixes broken link

dipikabh commented:
The `CSSRef` macro flaws showing up in this PR are similar to the `FirefixSidebar` macro flaws in https://github.com/mdn/content/pull/33450#issuecomment-2096752470, but at least the "Check URL issues" test is passing in other PR.

hamishwillee commented:
@dipikabh I'm unmarking myself as reviewer, because @estelle has far more background. If you need me to though, then please add me back.

dipikabh committed:
Merge branch 'main' into css-shape-function

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+- If the two shapes are of the same type, namely `ellipse()` and `circle()` and if their radii do not use the `closest-side` or `farthest-side` keywords, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `inset()`, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `polygon()`, and both polygons have the same number of vertices and use the same `<fill-rule>`, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `path()`, and both path strings have the same number and types of path data commands and in the same order, each path data command is interpolated as a real number.
```
```suggestion
- If both shapes are of type `path()`, and both path strings have the same number and types of [path data commands](/en-US/docs/Web/SVG/Attribute/d#path_commands) in the same order, each path data command parameter is interpolated as a real number.
```
the commands are the letters. The parameters are the numbers.

Side note: should we move https://developer.mozilla.org/en-US/docs/Web/CSS/path to be at https://developer.mozilla.org/en-US/docs/Web/CSS/basic-shape/path?

dipikabh replied:
- Updating it to:

> - **Both shapes are of type `path()`**: If the path strings in both have the same number and type of [path data commands](/en-US/docs/Web/SVG/Attribute/d#path_commands) in the same sequence, then each parameter of the path data command is interpolated as a real number.

- Side note
> should we move https://developer.mozilla.org/en-US/docs/Web/CSS/path to be at https://developer.mozilla.org/en-US/docs/Web/CSS/basic-shape/path?

Yes, thanks for saying this! I have been thinking the same. It is confusing to not have it under `<basic-shape>`. Also, the [spec](https://drafts.csswg.org/css-shapes-1/#supported-basic-shapes) lists each `<basic-shape>` function as a type; so, for example, the `inset()` function is really a parameter in the `<inset()>` type:

```
<inset()> = inset( <length-percentage>{1,4} [ round <'border-radius'> ]?)
```

I need to make these changes as well. I'll push out these changes shortly.

dipikabh replied:
Moved path() under basic-shape in https://github.com/mdn/content/pull/33446/commits/a50e5bd496a46d2b1293ad1c364364fb73b0f4b1

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+- If both shapes are of type `inset()`, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `polygon()`, and both polygons have the same number of vertices and use the same `<fill-rule>`, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `path()`, and both path strings have the same number and types of path data commands and in the same order, each path data command is interpolated as a real number.
+- If the two shapes are of type `shape()` and `path()` and have matching path data commands in length and order, the interpolated value is a `path()` function if both starting and ending values are `path()` functions; otherwise, it's treated as a `shape()` function. The interpolated value must represent the same list of path data commands, with numerical components interpolated between the corres ...
+- If both shapes are of type `shape()`, the shape commands are considered the same if they share the identical command keyword and `<by-to>` parameters; for `<curve-command>` and `<smooth-command>`, the number of control points must match.
```
```suggestion
- If both shapes are of type `shape()`, the shape commands are considered the same if they share the identical [shape command](/en-US/docs/Web/CSS/basic-shape/shape#shape-command) keywords and the same `by` and `to` parameters, in the same order. For `<curve-command>` and `<smooth-command>`, the number of control points must match.
- If one shape is of type `shape()` and the other is of type `path()` and both have matching path data commands in length and order, the shapes are interpolated as a `path()` function if both starting and ending values are `path()` functions; otherwise, they are interpolated as a `shape()` function. The interpolated value must represent the same list of path data commands, with numerical components interpolated between the corresponding components of the starting and ending list.

```
This is the start of the suggestion:
* reverse the order so 2 shapes comes before 1 shape and 1 path.
* update the text per the suggestion so it's clearer that we're talking about one of each.
* The text "The interpolated value must represent the same list of path data commands, with numerical components interpolated between the corresponding components of the starting and ending list." sounds like a directive to the browser vendor. Not sure it needs to be included, but if it does, rework it for what web developers need to know when animating between a shape and path.

dipikabh replied:
> reverse the order so 2 shapes comes before 1 shape and 1 path.

Agree, done

> update the text per the suggestion so it's clearer that we're talking about one of each.

Done

> rework it for what web developers need to know when animating between a shape and path

Updating it to:

> - **One shape is of type `path()` and the other is of type `shape()`**: Interpolation is applied between each corresponding value if the list of path data commands is identical in number as well as sequence. The interpolated shape is a `shape()` function, maintaining the same list of path data commands.

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+- If both shapes are of type `path()`, and both path strings have the same number and types of path data commands and in the same order, each path data command is interpolated as a real number.
+- If the two shapes are of type `shape()` and `path()` and have matching path data commands in length and order, the interpolated value is a `path()` function if both starting and ending values are `path()` functions; otherwise, it's treated as a `shape()` function. The interpolated value must represent the same list of path data commands, with numerical components interpolated between the corres ...
+- If both shapes are of type `shape()`, the shape commands are considered the same if they share the identical command keyword and `<by-to>` parameters; for `<curve-command>` and `<smooth-command>`, the number of control points must match.
+- If both shapes of type `shape()` with an `<arc-command>` have different `<arc-sweep>` between their starting and ending lists, then the interpolated result uses `cw` for any progress value between `0` and `1`. If the shapes have different `<arc-size>` keywords, then the interpolated result uses `large` for any progress value between `0` and `1`.
```
```suggestion
- If two `shape()` functions each have an `<arc-command>` with different `<arc-sweep>` directions, then the interpolated result goes clockwise, using the `cw` direction. If the shapes have different `<arc-size>` keywords, then the size is interpolated using the `large` value.
```
if all progress values are between 0 and 1, we can omit that part. The shape() page does not include the word 'progress'. I don't know what a progress value is exactly. Tried to make a suggestion that readers can understand on first read without going to the shape() page.

dipikabh replied:
This is great, thanks!

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+- If the two shapes are of type `shape()` and `path()` and have matching path data commands in length and order, the interpolated value is a `path()` function if both starting and ending values are `path()` functions; otherwise, it's treated as a `shape()` function. The interpolated value must represent the same list of path data commands, with numerical components interpolated between the corres ...
+- If both shapes are of type `shape()`, the shape commands are considered the same if they share the identical command keyword and `<by-to>` parameters; for `<curve-command>` and `<smooth-command>`, the number of control points must match.
+- If both shapes of type `shape()` with an `<arc-command>` have different `<arc-sweep>` between their starting and ending lists, then the interpolated result uses `cw` for any progress value between `0` and `1`. If the shapes have different `<arc-size>` keywords, then the interpolated result uses `large` for any progress value between `0` and `1`.
+- In all other cases, no interpolation occurs.
```
```suggestion
- In all other cases, no interpolation occurs and the animation is discrete.
```

estelle reviewed (dismissed):
review of the data-type page.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -107,7 +107,19 @@ The following shapes are supported. All `<basic-shape>` values use functional no
 
     The optional `<fill-rule>` represents the {{SVGAttr("fill-rule")}} used to determine the interior of the path. Possible values are `nonzero` and `evenodd`. Default value when omitted is `nonzero`.
```
Done. Retained the last line though about the default value when parameter is omitted.

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,276 @@
+
+    - `<hv-line-command>`: Specified as `[hline | vline] [by | to] <length-percentage>`. This command adds a horizontal (`hline`) or vertical (`vline`) [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. It draws a horizontal or vertical line from the command's starting point to its ending point. The `by`or `to` keyword determine the "relative" or ...
+
+    - `<curve-command>`: Specified as `curve [by | to] <coordinate-pair> via <coordinate-pair> [<coordinate-pair>]`. This command adds a [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve_commands) to the list of shape commands. The `by`or `to` keyword determine whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", r ...
```
Nice hint! Thanks

dipikabh replied:
The link on `d` page is working fine with the anchor as `#cubic_bézier_curve_commands` but when I copy paste from the URL, it does give me https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/SVG/Attribute/d#cubic_b%C3%A9zier_curve_commands.

With the suggested update for the link on the `shape()` page, the link is working in local preview. Hopefully it will work.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,276 @@
+
+      The `via` keyword specifies the control points of the curve as one or two `<coordinate-pair>` values. If a single `<coordinate-pair>` is provided, the command draws a [quadratic Bézier curve](/en-US/docs/Web/SVG/Attribute/d#quadratic-bézier-curve-commands); if two `<coordinate-pair>`s are provided, the command draws a cubic Bézier curve.
+
+    - `<smooth-command>`: Specified as `smooth [by | to] <coordinate-pair> [via <coordinate-pair>]`. This command adds a smooth [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve_commands) to the list of shape commands. The `by`or `to` keyword determine whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respective ...
```
Let me know if this works ;)

dipikabh replied:
Working in local preview

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
 ### Interpolation of basic shapes
 
-When animating between one `<basic-shape>` and another, the rules below are applied. The values in the shape functions {{Glossary("interpolation", "interpolate")}} as a simple list. The list values interpolate as {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If list values are not one of those types but are identical, those valu ...
+When animating between one `<basic-shape>` and another, the rules below are applied. The values in the `<basic-shape>` functions {{Glossary("interpolation", "interpolate")}} as a simple list. The list values interpolate as {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If list values are not one of those types but are identical,  ...
```
- I've updated the text slightly to say ("from one...to another" instead of "between" and "by computed value" instead of "as a simple list"):

> When animating from one `<basic-shape>` to another, the rules listed below are applied. The values in the `<basic-shape>` functions {{Glossary("interpolation", "interpolate")}} by computed value as a simple list. The list values interpolate as {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If list values are not one of those types but are identical, those values do interpolate.

Source: https://drafts.csswg.org/css-shapes-1/#basic-shape-interpolation

- About `calc()`, I'll need to check before I remove it. It is present in the interpolation for the newest `shape()` function: [wpt code for clip-path-interpolation-shape](https://github.com/web-platform-tests/wpt/blob/master/css/css-masking/animations/clip-path-interpolation-shape.html#L56-L66)

- Let me get back on these changes after I've checked with the Firefox engineer.

dipikabh replied:
Update after some clarifications from the engineer: I've added `<number>` and `<angle>` to the list.
Keeping `<calc>` around since it is not incorrect (and mentioned in the spec) and `<length-percentage>` accept `<calc>`. If values in both shapes are a combination of `<length>` and `<percentage>`, they are interpolated as `calc()`.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
-- If both shapes are of type `polygon()`, both polygons have the same number of vertices, and use the same `<fill-rule>`, interpolate between each value in the shape functions.
-- If both shapes are of type `path()`, both paths strings have the same number and types of path data commands in the same order, interpolate each path data command as real numbers.
-- In all other cases no interpolation occurs.
+- If the two shapes are of the same type, namely `ellipse()` and `circle()` and if their radii do not use the `closest-side` or `farthest-side` keywords, interpolation is applied between each value in the shape functions.
```
Updated to:

> - If the two shapes are of the same type, specifically `ellipse()` and `circle()`, and their radii do not use the `closest-side` or `farthest-side` keywords, then interpolation is applied between each corresponding value in those shape functions.

Does this work?

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
-- In all other cases no interpolation occurs.
+- If the two shapes are of the same type, namely `ellipse()` and `circle()` and if their radii do not use the `closest-side` or `farthest-side` keywords, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `inset()`, interpolation is applied between each value in the shape functions.
+- If both shapes are of type `polygon()`, and both polygons have the same number of vertices and use the same `<fill-rule>`, interpolation is applied between each value in the shape functions.
```
Updated this to (also changing the style of the list items slightly to aid in scanning the list visually):

> - Both shapes are of type `polygon()`: If they have the same number of vertices (specified as x/y coordinate pairs) and use the same `<fill-rule>`, then interpolation is applied between each corresponding value, including the coordinate pairs.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +146,17 @@ The values in a `<basic-shape>` function are computed as specified, with these e
-- If both shapes are of type `polygon()`, both polygons have the same number of vertices, and use the same `<fill-rule>`, interpolate between each value in the shape functions.
-- If both shapes are of type `path()`, both paths strings have the same number and types of path data commands in the same order, interpolate each path data command as real numbers.
-- In all other cases no interpolation occurs.
+- If the two shapes are of the same type, namely `ellipse()` and `circle()` and if their radii do not use the `closest-side` or `farthest-side` keywords, interpolation is applied between each value in the shape functions.
```
The two shapes need to be the same. Addressed in previous comment.

dipikabh commented:
I've added the following note to the `<basic-shape>`, `path()`, and `shape()` pages:

> **Note:** `<fill-rule>` is not supported in the {{cssxref("offset-path")}} property.

`path()` and `shape()` can both be used with `clip-path` and `offset-path` properties, so it is important to let folks know that `fill-rule` works (currently) only with the `clip-path` property.

dipikabh committed:
partial review fixes

dipikabh committed:
fixes anchor links

dipikabh committed:
another attempt at anchor links

dipikabh committed:
check if removing bezier works

dipikabh committed:
adjusts interpolation text

dipikabh committed:
undo the section title changes in d attribute

dipikabh committed:
moves path under basic-shape

dipikabh committed:
improves the basic-shape page

dipikabh committed:
updates links to path()

dipikabh committed:
missed this file in previous round

dipikabh commented:
Hi @estelle, thanks a lot for the review. I've addressed your feedback comments. Please note:

- I have moved `path()` under `<basic-shape>`. Because of this update, you'll see 1 new file added to this PR (Firefox release 97)
- I've revamped some part of the https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape page:
   - Some of the stuff from "Syntax" has been moved to "Description".
   - The function descriptions in "Syntax" are now as H3s subsections and common parameters are explained upfront.
   - I've revamped the [Interpolation](https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape#interpolation_of_basic_shapes) section. Hope it is clearer now.


(Sorry for the spam from the failing url test. It's fixed now.)

estelle committed:
Merge branch 'main' into css-shape-function

estelle commented:
> The `CSSRef` macro flaws showing up in this PR are similar to the `FirefixSidebar` macro flaws in [#33450 (comment)](https://github.com/mdn/content/pull/33450#issuecomment-2096752470), but at least the "Check URL issues" test is passing in other PR.

@OnkarRuikar fixed this in https://github.com/mdn/yari/pull/11081. updating branch in hopes of resolving. The PR has been merged, but I think yari might be on a random release cycle, so fingers crossed.

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+- `<fill-rule>`
 
-    A pair of insets in either dimension that add up to more than the used dimension (such as left and right insets of 75% apiece) define a shape enclosing no area. For this specification, this results in an empty float area.
+  - : Represents the {{SVGAttr("fill-rule")}} used to determine the interior of the shape. Possible values are `nonzero` and `evenodd`. Default value when omitted is `nonzero`. This parameter is specified in the `polygon()`, `path()`, and `shape()` functions.
```
```suggestion
  - : Sets the {{SVGAttr("fill-rule")}}, or filling rule, used to determine how the interior of the shape defined by the `polygon()`, `path()`, and `shape()` functions is to be filled.  Possible values are `nonzero` (the default) and `evenodd`.
```

dipikabh replied:
Fixed with a slight variation, links point to sub-sections in "Syntax":

  _- : Sets the {{SVGAttr("fill-rule")}} that is used to determine how the interior of the shape defined by the basic shapes [polygon](#polygon), [path](#path), and [shape](#shape-experimental_inline) is to be filled. Possible values are `nonzero` (the default) and `evenodd`._

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+  - : Represents the {{SVGAttr("fill-rule")}} used to determine the interior of the shape. Possible values are `nonzero` and `evenodd`. Default value when omitted is `nonzero`. This parameter is specified in the `polygon()`, `path()`, and `shape()` functions.
 
-- `{{cssxref("basic-shape/rect","rect()")}}`
+    > **Note:** `<fill-rule>` is not supported in the {{cssxref("offset-path")}} property.
```
Does it invalidate the entire declaration, or is it just ignored if included. If ignored, I maybe this:

```suggestion
    > **Note:** `<fill-rule>` is supported in the CSS  {{cssxref("offset-path")}} property and SVG `d` attribute but is ignored when passed as a parameter of a shape function within a {{cssxref("offset-path")}} property value.
```

dipikabh replied:
The presence of a `fill-rule` value invalidates the whole declaration:

<img src="https://github.com/mdn/content/assets/23019147/d684c78b-b3a6-40ea-a9a5-0b02757ffda6" alt="offset-path with evenodd fill-rule" />

<img src="https://github.com/mdn/content/assets/23019147/88cdcc3d-e3aa-4572-9098-401f6123d384" alt="offset-path with nonzero fill-rule" />

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+    > **Note:** `<fill-rule>` is not supported in the {{cssxref("offset-path")}} property.
 
-  - : Defines a rectangle using the specified distances from the top and left edges of the reference box.
+### Inset
```
We need to add some headings or an intro. Inset seems like a non-sequitor to 

A possible solution would be divide the syntax into two sections: "Common parameters" and "basic shape functions". 
```suggestion
### Basic shape functions

There are seven basic shape functions.

#### Rectangles by container inset
```

dipikabh replied:
Point taken. However:
- I do want to retain the shape sub-sections as H3s (to aid in navigation whenever we get support for H3s in page TOC) as against having them as bullet points or as H4s.
- Since the shape sections are sub-sections in "Syntax", perhaps qualifying the titles with "Syntax for" might add more clarity.

What do you think about:

```
## Syntax
### Common parameters
### Syntax for rectangles by container insets
### Syntax for rectangles by distance
### Syntax for rectangles with dimensions
### Syntax for circles
### Syntax for ellipses
### Syntax for polygons
### Syntax for paths
### Syntax for shapes {{Experimental_Inline}}
```

Additionally, I've updated the "Syntax" intro to:

_The `<basic-shape>` data type is used to create basic shapes including [rectangles by container insets](#syntax_for_rectangles_by_container_insets), [rectangles by distance](#syntax_for_rectangles_by_distance), [rectangles with dimensions](#syntax_for_rectangles_with_dimensions), [circles](#syntax_for_circles), [ellipses](#syntax_for_ellipses), [polygons](#syntax_for_polygons), [paths](#syntax_for_paths), and [shapes](#syntax_for_shapes_experimental_inline). These basic shapes are defined using one of the basic shape functions. All `<basic_shape>` values are CSS functions, with each value requiring a parameter that follows the shape's function-specific syntax._

estelle commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
-    ```css
-    rect( [ <length-percentage> | auto ]{4} [ round <`border-radius`> ]? )
-    ```
+The {{cssxref("basic-shape/inset","inset()")}} function defines an inset rectangle.
````
```suggestion
The {{cssxref("basic-shape/inset","inset()")}} function defines an inset rectangle with the size defined by the offset distance of each side from its container and, optionally, rounded corners.
```

estelle reviewed (commented):
reviewed more, but not done.

OnkarRuikar commented:
> @OnkarRuikar fixed this in [mdn/yari#11081](https://github.com/mdn/yari/pull/11081). updating branch in hopes of resolving. The PR has been merged, but I think yari might be on a random release cycle, so fingers crossed.

Uncross them fingers. The fix was released in yari `v2.51.1` but that version hasn't been published to [npmjs.com](https://www.npmjs.com/package/@mdn/yari?activeTab=versions). I don't remember if it's a manual publish or they've put some automation for it. cc/ @bsmth

dipikabh commented on files/en-us/web/css/basic-shape/path/index.md:
```diff
@@ -22,13 +22,16 @@ path(<string>)
 
 - [`<fill-rule>`](/en-US/docs/Web/SVG/Attribute/fill-rule) {{optional_inline}}
+
   - : An optional value of [`nonzero`](/en-US/docs/Web/SVG/Attribute/fill-rule#nonzero) (the default when omitted) or [`evenodd`](/en-US/docs/Web/SVG/Attribute/fill-rule#evenodd), which specifies the filling rule.
```
Nice! To make it more straightforward, updated as:

> - : An optional value of [`nonzero`](/en-US/docs/Web/SVG/Attribute/fill-rule#nonzero) (the default when omitted) or [`evenodd`](/en-US/docs/Web/SVG/Attribute/fill-rule#evenodd), which defines how the inside of the shape to be filled is determined.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
 ## Syntax
 
-The `<basic-shape>` data type is defined with one of the basic shape functions listed below.
+The `<basic-shape>` data type is defined using one of the basic shape functions described in the sections below. All `<basic-shape>` values use functional notation and are defined using the [value definition syntax](/en-US/docs/Web/CSS/Value_definition_syntax).
```
Got it, using your suggestion as is.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+- `round <'border-radius'>`
 
-![An image showing a circle inspected with the Firefox DevTools Shape Inspector. The different parts of the box model are highlighted.](shapes-reference-box.png)
+  - : Defines rounded corners for the inset rectangle using the same syntax as the CSS [`border-radius`](/en-US/docs/Web/CSS/border-radius) shorthand property. This parameter is specified in the `inset()`, `rect()`, and `xywh()` functions.
```
Done

dipikabh replied:
I got rid of the last sentence as per https://github.com/mdn/content/pull/33446#discussion_r1597739271

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+- `<shape-radius>`
 
-The following shapes are supported. All `<basic-shape>` values use functional notation and are defined here using the [value definition syntax](/en-US/docs/Web/CSS/Value_definition_syntax).
+  - : Defines the radius for a circle or an ellipse. The valid values include `<length>`, `<percentage>`, `closest-side` and `farthest-side`. Negative values are invalid. It defaults to `closest-side` if omitted.
```
> - link first occurrence of length and percentage.

Done

> - other dts code their shapes

What I've now done is added links to the shape sections on the page. For example,
  - : Defines rounded corners for [insets](#inset) and rectangles ([rectangle by distance](#rectangle_by_distance) and [rectangle with dimensions](#rectangle_with_dimensions)) using the same syntax as the CSS [`border-radius`](/en-US/docs/Web/CSS/border-radius) shorthand property.

and 

  - : Defines the radius for a [circle](#circle) or an [ellipse](#ellipse). Valid values include {{cssxref("length")}}, {{cssxref("percentage")}}, `closest-side` (the default), and `farthest-side`. Negative values are invalid.

> - Optionally, reduce verbiage by moving default to parens.

Done

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+    `closest-side` uses the length from the center of the shape to the closest side of the reference box. For circles, this is the closest side in any dimension. For ellipses, this is the closest side in the radius dimension.
 
-  - : Defines an inset rectangle.
+    `farthest-side` uses the length from the center of the shape to the farthest side of the reference box. For circles, this is the farthest side in any dimension. For ellipses, this is the farthest side in the radius dimension.
```
Done

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+- `<position>`
 
-    When all of the first four arguments are supplied, they represent the top, right, bottom and left offsets from the reference box inward that define the position of the edges of the inset rectangle. These arguments follow the syntax of the {{cssxref("margin")}} shorthand, which lets you set all four insets with one, two, three, or four values.
+  - : Defines the center of the shape. It defaults to center of the shape if omitted. This parameter is specified in the `circle()` and `ellipse()` functions.
```
Thanks! Updated as:

 _- : Defines the center [`<position>`](/en-US/docs/Web/CSS/position_value) of a [circle](#circle) or an [ellipse](#ellipse). It defaults to `center` if omitted._

dipikabh committed:
fixes review feedback

dipikabh commented on files/en-us/web/css/basic-shape/path/index.md:
```diff
@@ -1,13 +1,13 @@
 {{CSSRef}}
 
-The **`path()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) accepts an SVG path string, and is used in [CSS Shapes](/en-US/docs/Web/CSS/CSS_shapes) and CSS Motion Path to enable a shape to be drawn.
+The **`path()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) accepts an SVG path string, and is used in [CSS Shapes](/en-US/docs/Web/CSS/CSS_shapes) and CSS Motion Path to enable a shape to be drawn. The `path()` function is one of the {{cssxref("&lt;basic-shape&gt;")}} data types.
```
- I've linked "SVG path" to "/en-US/docs/Web/SVG/Element/path" instead

- Agree with the other comments. Updating with some adjustments:

_The **`path()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) accepts an [SVG path](/en-US/docs/Web/SVG/Element/path) string, and is used in the [CSS shapes](/en-US/docs/Web/CSS/CSS_shapes) and [CSS motion path](/en-US/docs/Web/CSS/CSS_motion_path) modules to enable a shape to be drawn. The `path()` function is a {{cssxref("&lt;basic-shape&gt;")}} data type value. It can be used in the CSS [`offset-path`](/en-US/docs/Web/CSS/offset-path) and [`clip-path`](/en-US/docs/Web/CSS/clip-path) properties and in the SVG [`d`](/en-US/docs/Web/SVG/Attribute/d) attribute._

(somewhat similar to what we have in https://developer.mozilla.org/en-US/docs/Web/CSS/basic-shape/rect)

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+- `round <'border-radius'>`
 
-![An image showing a circle inspected with the Firefox DevTools Shape Inspector. The different parts of the box model are highlighted.](shapes-reference-box.png)
+  - : Defines rounded corners for the inset rectangle using the same syntax as the CSS [`border-radius`](/en-US/docs/Web/CSS/border-radius) shorthand property. This parameter is specified in the `inset()`, `rect()`, and `xywh()` functions.
```
I got rid of the last sentence as per https://github.com/mdn/content/pull/33446#discussion_r1597739271

dipikabh committed:
fixes review feedback in basic-shape/path

dipikabh commented:
Thanks, @estelle, for your thorough review and helpful suggestions. I appreciate the quick review of the updated content and have addressed the comments so far.

Could I request you to prioritize the review of the new page https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape? A link to this page will be included in the Firefox release notes https://github.com/mdn/content/pull/33450/files (releasing Tuesday). Thank you!

estelle commented:
> Could I request you to prioritize the review of the new page https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape? A link to this page will be included in the Firefox release notes https://github.com/mdn/content/pull/33450/files (releasing Tuesday). Thank you!

I have meetings until about 1:00pm your time (10:00am mine) on Monday, but will try to review this after that.

bsmth commented:
> The fix was released in yari `v2.51.1` but that version hasn't been published to [npmjs.com](https://www.npmjs.com/package/@mdn/yari?activeTab=versions). I don't remember if it's a manual publish or they've put some automation for it. cc/ @bsmth


I believe we're waiting on https://github.com/mdn/yari/pull/11060 to be merged for the release to happen.

estelle reviewed (commented):
This is a review of all the files OTHER than shape and basic shape. Those are good to go.Will do shape and basic shape next

dipikabh commented on files/en-us/web/css/basic-shape/path/index.md:
```diff
@@ -22,13 +22,16 @@ path(<string>)
-  - : An optional value of [`nonzero`](/en-US/docs/Web/SVG/Attribute/fill-rule#nonzero) (the default when omitted) or [`evenodd`](/en-US/docs/Web/SVG/Attribute/fill-rule#evenodd), which specifies the filling rule.
+
+  - : An optional value of [`nonzero`](/en-US/docs/Web/SVG/Attribute/fill-rule#nonzero) (the default when omitted) or [`evenodd`](/en-US/docs/Web/SVG/Attribute/fill-rule#evenodd), which defines how the inside of the shape to be filled is determined.
+    > **Note:** `<fill-rule>` is not supported in the {{cssxref("offset-path")}} property.
```
Going with:

 > **Note:** `<fill-rule>` is not supported and invalid in {{cssxref("offset-path")}} and using it invalidates the property.

dipikabh committed:
updates the note text for <fill-rule>

estelle committed:
Merge branch 'main' into css-shape-function

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+The default reference box is the `margin-box`, as demonstrated in the below image which shows a circle created using `shape-outside: circle(50%)`. The shape is being defined with reference to the margin box.
 
-## Description
+![An image showing a circle inspected with the Firefox DevTools Shape Inspector. The different parts of the box model are highlighted.](shapes-reference-box.png)
```
sighted users may not know that's a dev tools view in line 126. we should add that, since otherwise what the image is demonstrating wont be clear to those who don't see the alt.

dipikabh replied:
Good point. Updating line 126 content to:

> The default reference box is the [`margin-box`](/en-US/docs/Web/CSS/box-edge#margin-box), as demonstrated in the image below. The image shows a circle created using `shape-outside: circle(50%)` and inspected using a browser's Developer Tools, highlighting the different parts of the box model. The shape here is defined with reference to the margin-box.

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +137,25 @@ The values in a `<basic-shape>` function are computed as specified, with these e
 ### Interpolation of basic shapes
 
-When animating between one `<basic-shape>` and another, the rules below are applied. The values in the shape functions {{Glossary("interpolation", "interpolate")}} as a simple list. The list values interpolate as {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If list values are not one of those types but are identical, those valu ...
+When animating from one `<basic-shape>` to another, the {{Glossary("interpolation")}} rules listed below are followed. For any interpolation to happen between two shapes, both must use the same reference box. The values between two `<basic-shape>` functions interpolate based on their computed values, forming a simple list. The list values are interpolated as {{cssxref("&lt;number&gt;")}}, {{cssxr ...
```
how would they use two different reference boxes?

unclear what a "simple list" is. Maybe a parameter list? Also, avoid the use of "simple"

estelle replied:
```suggestion
When animating from one `<basic-shape>` to another, the {{Glossary("interpolation")}} rules listed below are followed. For any interpolation to happen between two shapes, both must use the same reference box. The values between two `<basic-shape>` functions interpolate based on their computed values, forming a simple list. The list values are interpolated as {{cssxref("&lt;number&gt;")}}, {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, {{cssxref("&lt;angle&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. For parameter values outside of those data types, if the value is the same in all the basic shape functions, such as `nonzero` in both the from and to `basic-shape` functions, those values also interpolate.
```
The fill-rule or round only occurs once per shape, so i don't think it has to be in the same spot.

dipikabh replied:
> how would they use two different reference boxes?

we say this in text: "For any interpolation to happen between two shapes, both must use the same reference box."

> avoid the use of "simple"

Done

> i don't think it has to be in the same spot.

Actually, `<fill-rule>` in any other position will invalidate the property. So it is kind of implicit that `nonzero` or `evenodd`, if present in the from and to shapes, will be in the same position.

The updated version looks like this:

> When animating from one `<basic-shape>` to another, the {{Glossary("interpolation")}} rules listed below are followed. For any interpolation to happen between two shapes, both must use the same reference box. The values between two `<basic-shape>` functions interpolate based on their computed values, forming a list. The values in the list are interpolated as {{cssxref("&lt;number&gt;")}}, {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, {{cssxref("&lt;angle&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If the values are not one of those data types but are identical between the two interpolating basic shape functions, such as `nonzero`, those values also interpolate.

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +137,25 @@ The values in a `<basic-shape>` function are computed as specified, with these e
-When animating between one `<basic-shape>` and another, the rules below are applied. The values in the shape functions {{Glossary("interpolation", "interpolate")}} as a simple list. The list values interpolate as {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If list values are not one of those types but are identical, those valu ...
+When animating from one `<basic-shape>` to another, the {{Glossary("interpolation")}} rules listed below are followed. For any interpolation to happen between two shapes, both must use the same reference box. The values between two `<basic-shape>` functions interpolate based on their computed values, forming a simple list. The list values are interpolated as {{cssxref("&lt;number&gt;")}}, {{cssxr ...
+
+- **Both shapes are of type `ellipse()` or type `circle()`**: Interpolation is applied between each corresponding value if their radii are specified as {{cssxref("&lt;length-percentage&gt;")}} (rather than keywords such as `closest-side` or `farthest-side`).
```
* we used "length" or "percentage", not "length-percentage" in the definitions, so should probably stick to two separate terms here too.
* maybe note interpolated as ellipse if one of each, with the circle radius duplicated to create an ellipse.

dipikabh replied:
> we used "length" or "percentage", not "length-percentage" in the definitions, so should probably stick to two separate terms here too.

Done

> maybe note interpolated as ellipse if one of each, with the circle radius duplicated to create an ellipse.

Both have to be the same type. `circle()` and `ellipse()` are not interpolated.

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +137,25 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+
+- **Both shapes are of type `inset()`**: Interpolation is applied between each corresponding value.
+
+- **Both shapes are of type `polygon()`**: Interpolation is applied between each corresponding value, including the vertices (specified as x/y coordinate pairs), if they have the same number of vertices and use the same `<fill-rule>`.
```
```suggestion
- **Both shapes are of type `polygon()`**: Interpolation is applied between each corresponding value, specified as x/y coordinate pairs, if they have the same number of comma-separated pairs and use the same `<fill-rule>`.
```

dipikabh replied:
Since the corresponding values include both fill-rule and coordinate pairs, updating this to:

> - **Both shapes are of type `polygon()`**: Interpolation is applied between each corresponding value if they use the same `<fill-rule>` and have the same number of comma-separated coordinate pairs.

Change the order of fill-rule and coordinate pairs and removed "vertices" as suggested

estelle commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +137,25 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+
+- **Both shapes are of type `path()`**: Interpolation is applied to each parameter as a {{cssxref("&lt;number&gt;")}} if the path strings in both the shapes match the number, type, and sequence of [path data commands](/en-US/docs/Web/SVG/Attribute/d#path_commands).
+
+- **Both shapes are of type `shape()`**: Interpolation is applied between each corresponding value if they have the identical command keyword and use the same `<by-to>` keyword. If `shape()` is used in the {{cssxref("clip-path")}} property, the two shapes interpolate if they also have the same `<fill-rule>`.
```
`d` too?

dipikabh replied:
`shape()` to `shape()` will interpolate only if they use the same commands, same `by` or `to` keyword, (and the same fill rule in case of `clip-path`). The first parameter `from <coordinate-pair>` is anyway mandatory.

This is demo'ed at https://github.com/web-platform-tests/wpt/blob/4a0aeb02d32180d424668776162afc6f85d47d6f/css/css-masking/animations/clip-path-interpolation-shape.html#L86

And a negative test can be seen at https://github.com/web-platform-tests/wpt/blob/4a0aeb02d32180d424668776162afc6f85d47d6f/css/css-masking/animations/clip-path-interpolation-shape.html#L80.

estelle reviewed (commented):
review of basic-shape

estelle commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+```
 
-    The optional `<fill-rule>` represents the {{SVGAttr("fill-rule")}} used to determine the interior of the path. Possible values are `nonzero` and `evenodd`. Default value when omitted is `nonzero`.
+The required `<string>` is an [SVG Path](/en-US/docs/Web/SVG/Attribute/d) string encompassed in quotes.
````
```suggestion
The required `<string>` is a single [SVG Path](/en-US/docs/Web/SVG/Attribute/d) string encompassed in quotes with numeric values representing pixel units.
```
More info we may want to include.
we should create an issue about the `path()` page. We don't necessarily need to include this here, but we should include it there:

> While the `path()` function allows reuse of the SVG path syntax to define more arbitrary shapes than allowed by more specialized shape functions, it requires writing a path as a single string (which is not compatible with, for example, building a path piecemeal with `var()` and inherits a number of limitations from SVG, such as implicitly only allowing the `px` unit.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+- `from <coordinate-pair>`
+
+  - : Defines the starting point for the first `<shape-command>` as a pair of coordinates that are measured from the top-left corner of the reference box. The coordinates are specified as a {{cssxref("&lt;length-percentage&gt;")}} value. Add a comma after this parameter.
```
```suggestion
  - : Defines the starting point for the first `<shape-command>` as a pair of coordinates that are measured from the top-left corner of the reference box. The coordinates are specified as space-separated `x y` {{cssxref("&lt;length-percentage&gt;")}} values representing the left offset and top offset, respectively. Percentage values are relative to the width or height of the element's reference box, respectively. Add a comma after this parameter.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
````diff
@@ -0,0 +1,278 @@
+
+## Syntax
+
+```css
````
I didn't test these. Can you double check they all work?

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
````diff
@@ -0,0 +1,278 @@
+
+```css
+/* <fill-rule> */
+clip-path: shape(nonzero from 0px 0px, line to 10px 10px);
````
```suggestion
clip-path: shape(nonzero from 0 0, line to 10px 10px);
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `nonzero`: A point is considered inside the shape if a ray drawn from the point crosses more left-to-right than right-to-left path segments, resulting in a non-zero count. This is the default value when `<fill-rule>` is omitted.
+
+    - `evenodd`: A point is considered inside the shape if a ray drawn from the point crosses an odd number of path segments.
```
```suggestion
    - `evenodd`: A point is considered inside the shape if a ray drawn from the point crosses an odd number of path segments (the last time it crossed into the shape it did not cross back out.)
```
this change is not necessary, but the SVG fill rules values make so little sense and i don't think it is ever explained, until the "nonzero" above that i find my little aside helpful as a corollary to that definition

dipikabh replied:
Adding:

> This means that for each time the ray enters the shape, it has not exited an equal number of times, indicating an odd count of entries without corresponding exits.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<line-command>`: Specified as `line [by | to] <coordinate-pair>`. This command adds a [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. It draws a straight line from the command's starting point to its ending point. The `by`or `to` keyword specify whether the ending point specified by `<coordinate-pair>` is "relative" or "absolute", respe ...
+
+    - `<hv-line-command>`: Specified as `[hline | vline] [by | to] <length-percentage>`. This command adds a horizontal (`hline`) or vertical (`vline`) [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. It draws a horizontal or vertical line from the command's starting point to its ending point. The `by`or `to` keyword determine the "relative" or ...
```
```suggestion
    - `<hv-line-command>`: Specified as `[hline | vline] [by | to] <length-percentage>`. This command adds a horizontal (`hline`) or vertical (`vline`) [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. With `hline`, a horizontal line is drawn from the command's starting point `to` or `by` the `x` position defined by the `<length-percentage>`. For `vline`, a vertical line  is drawn from the command's starting point `to` or `by` the `y` position defined by the `<length-percentage>`. The `by` or `to` keyword determines the "relative" or "absolute" ending point, respectively. This command is equivalent to `<line-command>` with one coordinate value remaining unchanged from its starting command.
```

dipikabh replied:
Nice one, thanks! Only removed the "the" before "`<length-percentage>`"

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<curve-command>`: Specified as `curve [by | to] <coordinate-pair> via <coordinate-pair> [<coordinate-pair>]`. This command adds a [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve) to the list of shape commands. The `by`or `to` keyword determines whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respectiv ...
+
+      The `via` keyword specifies the control points of the curve as one or two `<coordinate-pair>` values. If a single `<coordinate-pair>` is provided, the command draws a [quadratic Bézier curve](/en-US/docs/Web/SVG/Attribute/d#quadratic_bézier_curve); if two `<coordinate-pair>`s are provided, the command draws a cubic Bézier curve.
```
```suggestion
      The `via` keyword specifies the control points of the cubic Bézier curve as one or two `<coordinate-pair>` values. If only a single `<coordinate-pair>` is provided, the command draws a [quadratic Bézier curve](/en-US/docs/Web/SVG/Attribute/d#quadratic_bézier_curve).
```
in both cases it's a cubic Bézier curve, so only iterate that a single value produces a special type of curve.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<smooth-command>`: Specified as `smooth [by | to] <coordinate-pair> [via <coordinate-pair>]`. This command adds a smooth [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve) to the list of shape commands. The `by`or `to` keyword determines whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respectively.
+
+      The `via` keyword can be optionally included to specify the control points of the curve through `<coordinate-pair>`. If `via <coordinate-pair>` is omitted, the command draws a smooth quadratic curve; if it is provided, the command draws a smooth cubic curve.
```
we need to say what the difference is between quadratic and smooth briefly in this page..

estelle replied:
ah, i already made the same comment

dipikabh replied:
I added an explanation

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+      > **Note:** If the starting and ending points of the arc lie on exactly opposite sides of the ellipse, there is only one possible ellipse and two possible arcs. In this case, `<arc-sweep>` specifies the arc to choose, and `<arc-size>` has no effect.
+
+    - `close`: Adds a [ClosePath command](/en-US/docs/Web/SVG/Attribute/d#closepath) to the list of shape commands.
```
```suggestion
    - `close`: Adds a [ClosePath command](/en-US/docs/Web/SVG/Attribute/d#closepath) to the list of shape commands, drawing a straight line from the current position (end of the last command) to the first point in the path defined in the `from x y,` parameter. To close the shape without drawing a line, use `moveto` with the originating coordinates. If the `closepath` command is immediately followed by a `move` command, the moveTo defines the starting point of the next subpath or shape. 
```

dipikabh replied:
Updated the names of the commands

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<arc-command>`: Specified as `arc [by | to] <coordinate-pair> of <length-percentage> [<length-percentage>] [<arc-sweep> | <arc-size> | rotate <angle>]`. This command adds an [elliptical arc curve command](/en-US/docs/Web/SVG/Attribute/d#elliptical_arc_curve) to the list of shape commands. It draws an elliptical arc between a starting point and an ending point. The `by`or `to` keyword deter ...
+
+      The elliptical arc curve command defines two possible ellipses, which intersect both the starting and ending points, and each can be traced in either direction, resulting in four possible arcs. The `of` keyword specifies the size of the ellipse fom which the arc is taken. The first `<length-percentage>` provides the horizontal radius of the ellipse, and the second provides the vertical radi ...
```
```suggestion
      The elliptical arc curve command defines two possible ellipses, which intersect both the starting and ending points, and each can be traced in either direction, resulting in four possible arcs depending on the arc size, direction, and angle. The `of` keyword specifies the size of the ellipse from which the arc is taken. The first `<length-percentage>` provides the horizontal radius of the ellipse, and the second provides the vertical radius. If only one `<length-percentage>` is provided, the same value is used for both radii. The following parameters help to determine which of the four arcs is used:
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    The following `<shape-command>`s can be specified:
+
+    - `<move-command>`: Specified as `move [by | to] <coordinate-pair>`. This command adds a [MoveTo command](/en-US/docs/Web/SVG/Attribute/d#moveto_path_commands) to the list of shape commands. It does not draw a shape; rather it specifies the position for the next command. The `by`or `to` keyword specify whether the `<coordinate-pair>` point is "relative" or "absolute", respectively.
```
```suggestion
    - `<move-command>`: Specified as `move [by | to] <coordinate-pair>`. This command adds a [MoveTo command](/en-US/docs/Web/SVG/Attribute/d#moveto_path_commands) to the list of shape commands. It draws nothing. Rather, it specifies the starting position for the next command. The `by`or `to` keyword specify whether the `<coordinate-pair>` point is "relative" or "absolute", respectively. If the `move` command follows a `closepath` command, it identifies the starting point of the next shape or subpath. 
```
shape felt like 'shape' rather than 'line', so making it clearer i hope

dipikabh replied:
Gotcha, yeah very nuanced, your version seems better to remove ambiguity.
Edited the last sentence for the names of these commands in `shape()`

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+However, `shape()` offers several advantages over using `path()`:
+
+- Unlike `path()`, which uses the [SVG path](/en-US/docs/Web/SVG/Element/path) syntax, `shape()` uses standard CSS syntax, making it easier to create and modify shapes directly in your stylesheet.
+- The `path()` function inherits some limitations from SVG, such as requiring shapes to be written as a single string and limiting units to `px`. The `shape()` function, on the other hand, supports a variety of CSS units, including percentages, `rem`, and `em`.
+- `shape()` also allows the use of CSS math functions, providing more versatility when defining shapes.
```
```suggestion
- `shape()` uses standard CSS syntax making it easier to create and modify shapes directly in your stylesheet than `path()`, which uses the [SVG path](/en-US/docs/Web/SVG/Element/path) syntax, .
- The `shape()` function supports a variety of CSS units, including percentages, `rem`, and `em`, unlike the `path()` function which defines shapes as a single string and limits units to `px`. 
- `shape()` also allows the use of CSS math functions, providing more versatility when defining shapes.
```
put shape first in all to see the benefits of current page as compared to drawbacks of something defined elsewhere.

dipikabh replied:
Good point. Fixed.

estelle reviewed (commented):
i got up to the examples part. I'll just the examples are good :D and call this read done.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+The `<basic-shape>` data type is used to create basic shapes including [rectangles by container insets](#syntax_for_rectangles_by_container_insets), [rectangles by distance](#syntax_for_rectangles_by_distance), [rectangles with dimensions](#syntax_for_rectangles_with_dimensions), [circles](#syntax_for_circles), [ellipses](#syntax_for_ellipses), [polygons](#syntax_for_polygons), [paths](#syntax_fo ...
 
-When creating a shape, the reference box is defined by each property that uses `<basic-shape>` values. The coordinate system for the shape has its origin at the top-left corner of the reference box, with the x-axis running to the right and the y-axis running downwards. All the lengths expressed in percentages are resolved from the used dimensions of the reference box.
+### Common parameters
```
Thank you!

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
-    ```css
-    rect( [ <length-percentage> | auto ]{4} [ round <`border-radius`> ]? )
-    ```
+The {{cssxref("basic-shape/inset","inset()")}} function defines an inset rectangle with its size defined by the offset distance of each side from its container and, optionally, rounded corners.
````
Good call! I made some changes to use "creates" and adding a comma

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
-    ```css
-    xywh( <length-percentage>{2} <length-percentage [0,∞]>{2} [ round <`border-radius`> ]? )
-    ```
+The {{cssxref("basic-shape/rect","rect()")}} function defines a rectangle using the specified distances from the top and left edges of the reference box.
````
Thanks, just moved the position of optional

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+```
 
-- `{{cssxref("basic-shape/circle","circle()")}}`
+You specify four values to create the rectangle. Each of the four values is either a {{cssxref("length")}}, a {{cssxref("percentage")}}, or the keyword `auto`. When using the `rect()` function, you do not define the width and height of the rectangle. The rectangle's dimensions depend on the size of the reference box and the offset values.
````
This is excellent, adding more clarity, so including it. Added some edits though.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
-    ```css
-    circle( <shape-radius>? [ at <position> ]? )
-    ```
+The {{cssxref("basic-shape/xywh","xywh()")}} function defines a rectangle using the specified distances from the top and left edges of the reference box and the specified width and height of the rectangle.
````
Accepted. Only amended the last "optional" phrase

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+ellipse( [ <shape-radius>{2} ]? [ at <position> ]? )
+```
+
+The `<shape-radius>` arguments represent rx and ry, the x-axis and y-axis radii of the ellipse, in that order. Percentage values here are resolved against the used width (for the rx value) and the used height (for the ry value) of the reference box.
````
Thank you

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+## Description
 
-`closest-side` uses the length from the center of the shape to the closest side of the reference box. For circles, this is the closest side in any dimension. For ellipses, this is the closest side in the radius dimension.
+When creating a shape, the reference box is defined by each property that uses `<basic-shape>` values. The coordinate system for the shape has its origin at the top-left corner of the reference box, with the x-axis running to the right and the y-axis running downwards. All the lengths expressed in percentages are resolved from the used dimensions of the reference box.
```
Good catch, thanks!
About "element's margin box", I have been using "element's reference box", but I'll defer to your expertise here

dipikabh committed:
Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

dipikabh committed:
Apply suggestions from linter

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dipikabh committed:
Apply suggestions from linter

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+```
 
-    Each pair argument in the list represents _xi_ and _yi_ - the x and y axis coordinates of the vertex of the polygon at position i.
+Each `<length-percentage>` pair in the list represents _xi_ and _yi_ - the x and y axis coordinates of the vertex of the polygon at position i.
````
Makes sense. Further edited to:

> The function takes a list of comma-separated coordinate pairs, each consisting of two space-separated `<length-percentage>` values as the _xi_ and _yi_ pair. These values represents the x and y axis coordinates of the vertex of the polygon at position _i_.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+## Description
 
-`closest-side` uses the length from the center of the shape to the closest side of the reference box. For circles, this is the closest side in any dimension. For ellipses, this is the closest side in the radius dimension.
+When creating a shape, the reference box is defined by the element that uses `<basic-shape>` values. The coordinate system for the shape has its origin at the top-left corner of the element's margin box by default, with the x-axis running to the right and the y-axis running downwards. All the lengths expressed in percentages are resolved from the dimensions of the reference box.
```
I was going through the [basic shape functions spec](https://drafts.csswg.org/css-shapes/#basic-shape-functions) again and this caught my attention: "the reference box is defined by each property that uses `<basic-shape>` values.". On second thoughts, this makes sense because the basic shape function is used as a value of a property and not directly by the element itself. The property (`clip-path` or `offset-path` for example) can define the reference box to be used. The `<basic-shape>` function will then work within that reference box.
So reverting the "element" to "property" here.
```suggestion
When creating a shape, the reference box is defined by the property that uses `<basic-shape>` values. The coordinate system for the shape has its origin at the top-left corner of the element's margin box by default, with the x-axis running to the right and the y-axis running downwards. All the lengths expressed in percentages are resolved from the dimensions of the reference box.
```

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+When creating a shape, the reference box is defined by each property that uses `<basic-shape>` values. The coordinate system for the shape has its origin at the top-left corner of the reference box, with the x-axis running to the right and the y-axis running downwards. All the lengths expressed in percentages are resolved from the used dimensions of the reference box.
 
-`farthest-side` uses the length from the center of the shape to the farthest side of the reference box. For circles, this is the farthest side in any dimension. For ellipses, this is the farthest side in the radius dimension.
+The default reference box is the `margin-box`, as demonstrated in the below image which shows a circle created using `shape-outside: circle(50%)`. The shape is being defined with reference to the margin box.
```
I think the reference box default can be changed, for example by specifying [`<geometry-box>`](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path#geometry-box) in `clip-path` and [`<coord-box>`](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-path#offset-path) in `offset-path`

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+{{CSSRef}}
+
+The **`shape()`** [CSS function](/en-US/docs/Web/CSS/CSS_Functions) is used to define a shape for the {{cssxref("clip-path")}} and {{cssxref("offset-path")}} properties. It combines an initial starting point with a series of shape commands that define the path of the shape. The `shape()` function is one of the {{cssxref("&lt;basic-shape&gt;")}} data types.
```
> just want to confirm, not d or shape-outside?

I believe so. Those are the only two properties, "clip-path" and "offset-path", for which we had Firefox bugs (listed in the PR description) and for which we have [web platform tests](https://github.com/search?q=repo%3Aweb-platform-tests%2Fwpt+https%3A%2F%2Fdrafts.csswg.org%2Fcss-shapes-2%2F%23shape-function&type=code).

> Do we want to add that it's similar to path except with CSS instead of SVG norms?

This info is covered in the "Description" section

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+- `from <coordinate-pair>`
+
+  - : Defines the starting point for the first `<shape-command>` as a pair of coordinates that are measured from the top-left corner of the reference box. The coordinates are specified as a {{cssxref("&lt;length-percentage&gt;")}} value. Add a comma after this parameter.
```
Done

dipikabh committed:
fixes shape and basic-shape feedback

dipikabh committed:
element -> property

dipikabh commented:
Hi @estelle, thanks again for the great feedback. I've taken care of all your comments except for explaining quadratic and smooth curves. I think we're close to landing this.

dipikabh commented:
Hi @estelle, could I get an approval on this PR? If there's any outstanding feedback, I can address them via a new issue/PR. Thank you!

estelle commented:
> Hi @estelle, could I get an approval on this PR? If there's any outstanding feedback, I can address them via a new issue/PR. Thank you!

I already created three new MDN/content/issues to address issues that don't need to be addressed to merge this PR.

The re-review of this PR is on my todo list.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<move-command>`: Specified as `move [by | to] <coordinate-pair>`. This command adds a [MoveTo command](/en-US/docs/Web/SVG/Attribute/d#moveto_path_commands) to the list of shape commands. It draws nothing. Rather, it specifies the starting position for the next command. The `by`or `to` keyword specify whether the `<coordinate-pair>` point is "relative" or "absolute", respectively. If the ` ...
+
+    - `<line-command>`: Specified as `line [by | to] <coordinate-pair>`. This command adds a [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. It draws a straight line from the command's starting point to its ending point. The `by`or `to` keyword specify whether the ending point specified by `<coordinate-pair>` is "relative" or "absolute", respe ...
```
```suggestion
    - `<line-command>`: Specified as `line [by | to] <coordinate-pair>`. This command adds a [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. It draws a straight line from the command's starting point to its ending point. The `by` or `to` keyword specifies whether the ending point specified by `<coordinate-pair>` is "relative" or "absolute", respectively.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    The following `<shape-command>`s can be specified:
+
+    - `<move-command>`: Specified as `move [by | to] <coordinate-pair>`. This command adds a [MoveTo command](/en-US/docs/Web/SVG/Attribute/d#moveto_path_commands) to the list of shape commands. It draws nothing. Rather, it specifies the starting position for the next command. The `by`or `to` keyword specify whether the `<coordinate-pair>` point is "relative" or "absolute", respectively. If the ` ...
```
"keyword specifies" or "keywords specify"

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+      The `via` keyword can be optionally included to specify the control points of the curve through `<coordinate-pair>`. If `via <coordinate-pair>` is omitted, the command draws a smooth quadratic curve; if it is provided, the command draws a smooth cubic curve.
+
+    - `<arc-command>`: Specified as `arc [by | to] <coordinate-pair> of <length-percentage> [<length-percentage>] [<arc-sweep> | <arc-size> | rotate <angle>]`. This command adds an [elliptical arc curve command](/en-US/docs/Web/SVG/Attribute/d#elliptical_arc_curve) to the list of shape commands. It draws an elliptical arc between a starting point and an ending point. The `by`or `to` keyword deter ...
```
```suggestion
    - `<arc-command>`: Specified as `arc [by | to] <coordinate-pair> of <length-percentage> [<length-percentage>] [<arc-sweep> | <arc-size> | rotate <angle>]`. This command adds an [elliptical arc curve command](/en-US/docs/Web/SVG/Attribute/d#elliptical_arc_curve) to the list of shape commands. It draws an elliptical arc between a starting point and an ending point. The `by` or `to` keyword determines whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respectively.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<arc-command>`: Specified as `arc [by | to] <coordinate-pair> of <length-percentage> [<length-percentage>] [<arc-sweep> | <arc-size> | rotate <angle>]`. This command adds an [elliptical arc curve command](/en-US/docs/Web/SVG/Attribute/d#elliptical_arc_curve) to the list of shape commands. It draws an elliptical arc between a starting point and an ending point. The `by`or `to` keyword deter ...
+
+      The elliptical arc curve command defines two possible ellipses, which intersect both the starting and ending points, and each can be traced in either direction, resulting in four possible arcs depending on the arc size, direction, and angle. The `of` keyword specifies the size of the ellipse from which the arc is taken. The first `<length-percentage>` provides the horizontal radius of the e ...
```
```suggestion
      The elliptical arc curve command defines two possible ellipses, which intersect both the starting and ending points, and each can be traced in either direction, resulting in four possible arcs depending on the arc size, direction, and angle. The `of` keyword specifies the size of the ellipse from which the arc is taken. The first `<length-percentage>` provides the horizontal radius of the ellipse and the second provides the vertical radius. If only one `<length-percentage>` is provided, the same value is used for both radii. The following parameters help to determine which of the four arcs are used:
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+      > **Note:** If the starting and ending points of the arc lie on exactly opposite sides of the ellipse, there is only one possible ellipse and two possible arcs. In this case, `<arc-sweep>` specifies the arc to choose, and `<arc-size>` has no effect.
+
+    - `close`: Adds a [ClosePath command](/en-US/docs/Web/SVG/Attribute/d#closepath) to the list of shape commands, drawing a straight line from the current position (end of the last command) to the first point in the path defined in the `from <coordinate-pair>` parameter. To close the shape without drawing a line, use `<move-command>` with the originating coordinates. If the `close` command is i ...
```
```suggestion
    - `close`: Adds a [ClosePath command](/en-US/docs/Web/SVG/Attribute/d#closepath) to the list of shape commands, drawing a straight line from the current position (end of the last command) to the first point in the path defined in the `from <coordinate-pair>` parameter. To close the shape without drawing a line, include a `<move-command>` with the originating coordinates before the close command. If the `close` command is immediately followed by a `<move-command>`, it defines the starting point of the next shape or subpath.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+## Description
+
+The `shape()` function allows you to define complex shapes. It is similar to the {{cssxref("basic-shape/path","path()")}} function in several ways:
```
```suggestion
The `shape()` function allows you to define complex shapes. It is similar to the {{cssxref("basic-shape/path","path()")}} shape function in several ways:
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+However, `shape()` offers several advantages over using `path()`:
+
+- `shape()` uses standard CSS syntax, making it easier to create and modify shapes directly in your stylesheet. In comparison, `path()` uses the [SVG path](/en-US/docs/Web/SVG/Element/path) syntax, which is less intuitive for those not familiar with SVG.
+- `shape()` supports a variety of CSS units, including percentages, `rem`, and `em`. `path()`, on the other hand, defines shapes as a single string and limits units to `px`.
```
missing path()

dipikabh replied:
Not sure what this comment means

estelle replied:
I think i meant `calc()`

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+- `shape()` uses standard CSS syntax, making it easier to create and modify shapes directly in your stylesheet. In comparison, `path()` uses the [SVG path](/en-US/docs/Web/SVG/Element/path) syntax, which is less intuitive for those not familiar with SVG.
+- `shape()` supports a variety of CSS units, including percentages, `rem`, and `em`. `path()`, on the other hand, defines shapes as a single string and limits units to `px`.
+- `shape()` also allows the use of CSS math functions, providing more versatility when defining shapes.
```
```suggestion
- `shape()` allows the use of CSS math functions, providing versatility when defining shapes.
```

estelle replied:
does this mean calc or other math functions? we should link or example this if its not just calc() (there's already a calc example above)

dipikabh replied:
I'll have to come back to this later outside this PR. The spec does say math functions but in code I see only calc() examples

estelle replied:
I tested it.
```suggestion
- `shape()` also allows the use of CSS math functions, like {{cssxref("calc")}}, {{cssxref("max")}}, and {{cssxref("abs")}}, providing more versatility when defining shapes.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+### Using `shape()` to define a path
+
+This example demonstrates how the `shape()` function can be used in the {{cssxref("offset-path")}} property to define the shape of the path that an element can follow. We're creating a curved path in this example.
```
```suggestion
This example demonstrates how the `shape()` function can be used in the {{cssxref("offset-path")}} property to define the shape of the path an element can follow. We're creating a curved path in this example.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+}
+
+.shape1 {
+  offset-path: shape(from 30% 60px, curve to 180px 180px via 90px 190px);
```
i see `close` isn't used. we should mention above that closing a shape is optional and maybe explain the point of closing a shape

dipikabh replied:
`close` is one of the shape commands and hasn't been flagged as mandatory in the shape command list. But you have a fair point about explaining in text what happens when we don't add `close`. I'm afraid I'll have to come back to this later (after this PR) along with info about `calc()` vs other math functions

estelle reviewed (commented):
review of shape() function page

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+- `<shape-command>`
+
+  - : Specifies one or a list of comma-seperated commands that define the shape, using syntax similar to [SVG path commands](/en-US/docs/Web/SVG/Attribute/d#path_commands). The commands to define a shape include `<move-command>`, `<line-command>`, `<hv-line-command>`, `<curve-command>`, `<smooth-command>`, `<arc-command>`, and `close`. Each command's starting point is the previous command's endin ...
```
Accepting all except the "with"

dipikabh committed:
Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<curve-command>`: Specified as `curve [by | to] <coordinate-pair> via <coordinate-pair> [<coordinate-pair>]`. This command adds a [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve) to the list of shape commands. The `by` or `to` keyword determines whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respecti ...
+
+      The `via` keyword specifies the control points of the cubic Bézier curve as one or two `<coordinate-pair>` values. If only a single `<coordinate-pair>` is provided, the command draws a [quadratic Bézier curve](/en-US/docs/Web/SVG/Attribute/d#quadratic_bézier_curve).
```
```suggestion
      The `via` keyword specifies the control points of the Bézier curve. If only a single `<coordinate-pair>` is provided, the command draws a [quadratic Bézier curve](/en-US/docs/Web/SVG/Attribute/d#quadratic_bézier_curve), which is defined by three points (the start point, control point, and end point). If two `<coordinate-pair>` values are provided, the command draws a cubic Bézier curve, which is defined by four points (the start point, two control points, and the end point).
```

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<smooth-command>`: Specified as `smooth [by | to] <coordinate-pair> [via <coordinate-pair>]`. This command adds a smooth [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve) to the list of shape commands. The `by` or `to` keyword determines whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respectively.
+
+      The `via` keyword can be optionally included to specify the control points of the curve through `<coordinate-pair>`. If `via <coordinate-pair>` is omitted, the command draws a smooth quadratic curve; if it is provided, the command draws a smooth cubic curve.
```
```suggestion
      The `via` keyword can be optionally included to specify the control points of the curve through `<coordinate-pair>`. If `via <coordinate-pair>` is omitted, the command draws a smooth quadratic Bézier curve, which uses the previous control point and the current end point to define the curve. If it is provided, the command draws a smooth cubic Bézier curve, which is defined by the previous control point, the current control point, and the current end point. Smooth curves ensure a continuous transition from the shape, while quadratic curves do not. Smooth quadratic curves maintain a seamless transition using a single control point, whereas smooth cubic curves provide a more refined transition using two control points.
```

dipikabh committed:
Apply suggestions from code review

dipikabh commented:
Hi @estelle, let me know if you have any more feedback for the shape() page. If not, could I get an approval on this PR? Thank you!

OnkarRuikar commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+  100% {
+    offset-distance: 100%;
+  }
+}
```
```suggestion
}

@supports not (clip-path: shape(from 0 0, move to 0 0)) {
  .container {
    display: none;
  }

  body::after {
    content: "Your browser doesn't support the feature yet.";
  }
}
```
Let's save readers some time before they spend time figuring out that it's not supported in their browsers yet.

estelle replied:
how about "Your browser doesn't support the shape() function yet.";

dipikabh replied:
Currently, `shape()` is implemented only in Firefox, behind a preference.

OnkarRuikar commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+.shape2 {
+  offset-path: shape(
+    from 20px 20px,
+    move to 50px 90px,
+    hline to 8em,
```
```suggestion
    from 50px 90px,
    hline to 8em,
    move to 50px 90px,
```
In your code the `move` command is being redundant. Because [the `move` command draws nothing](https://drafts.csswg.org/css-shapes-2/#typedef-shape-move-command), it merely moves the pointer thus it is not animaable.

If we simply move the `move to` command after `hline to` we can infact demo this fact.

dipikabh replied:
Done, although I've also updated CSS for `shape1` to demo `move`

OnkarRuikar commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+    curve by -20px 60% via 10% 40px 20% 20px,
+    smooth by -40% -10px via -10px 70px
+  );
+}
```
```suggestion
}

@supports not (clip-path: shape(from 0 0, move to 0 0)) {
  .container {
    display: none;
  }

  body::after {
    content: "Your browser doesn't support the feature yet.";
  }
}
```

OnkarRuikar reviewed (commented):
Looking at the cool animation demos I couldn't help myself.

estelle reviewed (commented):
still working on it.

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
````diff
@@ -0,0 +1,278 @@
+
+```css
+/* <fill-rule> */
+clip-path: shape(nonzero from 0 0, line to 10px 10px);
````
The intention here is to show only one `<shape-command>`

dipikabh committed:
Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

dipikabh commented:
Hi @estelle, I see more comments on the basic-shape page when the focus of this PR is to document the `shape()` function. I would appreciate it if you could limit your feedback to the `shape()` function in this PR. I'll take up your review comments on the basic-shape page in a separate PR.

> still working on it.

Let me know when you've completed the review of the `shape()` function page so that I can address all comments in one go.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<line-command>`: Specified as `line [by | to] <coordinate-pair>`. This command adds a [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. It draws a straight line from the command's starting point to its ending point. The `by` or `to` keyword specifies whether the ending point specified by `<coordinate-pair>` is "relative" or "absolute", re ...
+
+    - `<hv-line-command>`: Specified as `[hline | vline] [by | to] <length-percentage>`. This command adds a horizontal (`hline`) or vertical (`vline`) [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. With `hline`, a horizontal line is drawn from the command's starting point `to` or `by` the `x` position defined by `<length-percentage>`. With ` ...
```
```suggestion
    - `<hv-line-command>`: Specified as `[hline | vline] [by | to] <length-percentage>`. This command adds a horizontal (`hline`) or vertical (`vline`) [LineTo command](/en-US/docs/Web/SVG/Attribute/d#lineto_path_commands) to the list of shape commands. With `hline`, a horizontal line is drawn from the command's starting point `to` or `by` the `x` position defined by `<length-percentage>`. With `vline`, a vertical line is drawn from the command's starting point `to` or `by` the `y` position defined by `<length-percentage>`. The `by` or `to` keyword determines the "relative" or "absolute" ending point, respectively. This command is equivalent to `<line-command>` with one coordinate value set by the single `<length-percentage>` and the other coordinate value remaining unchanged from its starting command.
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<smooth-command>`: Specified as `smooth [by | to] <coordinate-pair> [via <coordinate-pair>]`. This command adds a smooth [Bézier curve command](/en-US/docs/Web/SVG/Attribute/d#cubic_bézier_curve) to the list of shape commands. The `by` or `to` keyword determines whether the ending point of the curve, specified by the first `<coordinate-pair>`, is "relative" or "absolute", respectively.
+
+      The `via` keyword can be optionally included to specify the control points of the curve through `<coordinate-pair>`. If `via <coordinate-pair>` is omitted, the command draws a smooth quadratic Bézier curve, which uses the previous control point and the current end point to define the curve. If it is provided, the command draws a smooth cubic Bézier curve, which is defined by the previous co ...
```
```suggestion
      If `via <coordinate-pair>` is omitted, the command draws a smooth quadratic Bézier curve, which uses the previous control point and the current endpoint to define the curve. If the optional `via` keyword is included, it specifies the control points of the curve through `<coordinate-pair>`, drawing a smooth cubic Bézier curve defined by the previous control point, the current control point, and the current endpoint. Smooth curves ensure a continuous transition from the shape, while quadratic curves do not. Smooth quadratic curves maintain a seamless transition using a single control point, whereas smooth cubic curves provide a more refined transition using two control points.
```
reordered b/c the 2nd sentence was separating the first sentence from its third sentence explanation. The first sentence introduces a non-intuitive concept, so want to keep it together with its explanation

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    - `<arc-command>`: Specified as `arc [by | to] <coordinate-pair> of <length-percentage> [<length-percentage>] [<arc-sweep> | <arc-size> | rotate <angle>]`. This command adds an [elliptical arc curve command](/en-US/docs/Web/SVG/Attribute/d#elliptical_arc_curve) to the list of shape commands. It draws an elliptical arc between a starting point and an ending point. The `by` or `to` keyword dete ...
+
+      The elliptical arc curve command defines two possible ellipses, which intersect both the starting and ending points, and each can be traced in either direction, resulting in four possible arcs depending on the arc size, direction, and angle. The `of` keyword specifies the size of the ellipse from which the arc is taken. The first `<length-percentage>` provides the horizontal radius of the e ...
```
```suggestion
      The elliptical arc curve command defines two possible ellipses, which intersect both the starting and ending points, and each can be traced clockwise or counterclockwise, resulting in four possible arcs depending on the arc size, direction, and angle. The `of` keyword specifies the size of the ellipse from which the arc is taken. The first `<length-percentage>` provides the horizontal radius of the ellipse and the second provides the vertical radius. If only one `<length-percentage>` is provided, the value is used for both radii (the radius of a circle). The following parameters help to determine which of the four arcs are used:
```

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+      - `<arc-size>`: Indicates whether the desired arc is the larger (`large`) or smaller (`small`) of the two arcs. If omitted, this defaults to `small`.
+      - `<angle>`: Specifies the angle, in degrees, by which to rotate the ellipse relative to the x-axis. A positive angle rotates the ellipse clockwise, and a negative angle rotates it counter-clockwise. If omitted, this defaults to `0deg`.
+
+      > **Note:** If the starting and ending points of the arc lie on exactly opposite sides of the ellipse, there is only one possible ellipse and two possible arcs. In this case, `<arc-sweep>` specifies the arc to choose, and `<arc-size>` has no effect.
```
if only one radius is included, does the arc-size value make a difference? or angle? Should we add a note about this, or add a sentence to the arc size and angle descriptions?

dipikabh replied:
Good point. Combined the various situations and listed as bullet points. 
    

> Special situations are handled as follows:
> 
> - If only one `<length-percentage>` is provided, the same value is used for both the horizontal and vertical radii, effectively creating a circle. In this case, `<arc-size>` and `<angle>` have no affect.
> - If either radius is zero, the command is equivalent to a `<line-command>` to the ending point.
> - If either radius is negative, its absolute value is used instead.
> - If the horizontal and vertical radii don’t describe an ellipse large enough to intersect both the starting point and ending points (after rotation by the specified `<angle>`), the radii are scaled up uniformly until the ellipse is just large enough to intersect both points.
> - If the starting and ending points of the arc lie on exactly opposite sides of the ellipse, there is only one possible ellipse and two possible arcs. In this case, `<arc-sweep>` specifies the arc to choose, and `<arc-size>` has no effect.

See `<arc-command>` in https://pr33783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape#parameters for other accompanying updates.

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+      > **Note:** If the starting and ending points of the arc lie on exactly opposite sides of the ellipse, there is only one possible ellipse and two possible arcs. In this case, `<arc-sweep>` specifies the arc to choose, and `<arc-size>` has no effect.
+
+    - `close`: Adds a [ClosePath command](/en-US/docs/Web/SVG/Attribute/d#closepath) to the list of shape commands, drawing a straight line from the current position (end of the last command) to the first point in the path defined in the `from <coordinate-pair>` parameter. To close the shape without drawing a line, include a `<move-command>` with the originating coordinates before the close comma ...
```
again, the "If the <move-command> follows the close command," part would definitely benefit from an example.

dipikabh replied:
Done as indicated in https://github.com/mdn/content/pull/33446#discussion_r1616444437

estelle commented on files/en-us/web/css/basic-shape/path/index.md:
```diff
@@ -22,13 +22,16 @@ path(<string>)
 - [`<fill-rule>`](/en-US/docs/Web/SVG/Attribute/fill-rule) {{optional_inline}}
-  - : An optional value of [`nonzero`](/en-US/docs/Web/SVG/Attribute/fill-rule#nonzero) (the default when omitted) or [`evenodd`](/en-US/docs/Web/SVG/Attribute/fill-rule#evenodd), which specifies the filling rule.
+
+  - : An optional value of [`nonzero`](/en-US/docs/Web/SVG/Attribute/fill-rule#nonzero) (the default when omitted) or [`evenodd`](/en-US/docs/Web/SVG/Attribute/fill-rule#evenodd), which defines how the inside of the shape to be filled is determined.
+    > **Note:** `<fill-rule>` is not supported in {{cssxref("offset-path")}} and using it invalidates the property.
```
```suggestion
  - : Defines how the inside of the shape created by the path is to be filled. The values include:
  
      - `nonzero`: A point is considered inside the shape if a ray drawn from the point crosses more left-to-right than right-to-left path segments, resulting in a non-zero count. This is the default value when `<fill-rule>` is omitted.

    - `evenodd`: A point is considered to be inside the shape if a ray drawn from the point crosses an odd number of path segments. This means that for each time the ray enters the shape, it has not exited an equal number of times, indicating an odd count of entries without corresponding exits.
    
    > **Warning:** `<fill-rule>` is not supported in {{cssxref("offset-path")}} and using it invalidates the property.
```

dipikabh replied:
Fixed in https://pr33783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/path#parameters

estelle commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+    curve to 60px 20% via 40px 0,
+    smooth to 90px 0,
+    curve by -20px 60% via 10% 40px 20% 20px,
+    smooth by -40% -10px via -10px 70px
```
```suggestion
    smooth by -40% -10px via -10px 70px,  close,
    move to 100px 100px,
    hline by 50px,
    vline by 50px,
    hline by -50px,
    close
```
if you want.
This shows the example of the "moveto" after a close, and adds a box.

dipikabh replied:
Done in https://pr33783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape#using_shape_to_define_the_visible_part_of_an_element. Also updated the example description.

estelle reviewed (commented):
reviewed shape

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+- `<shape-command>`
+
+  - : Specifies a list of one or more comma-separated commands that define the shape, using syntax similar to [SVG path commands](/en-US/docs/Web/SVG/Attribute/d#path_commands). Commands include `<move-command>`, `<line-command>`, `<hv-line-command>`, `<curve-command>`, `<smooth-command>`, `<arc-command>`, and `close`. Each command's starting point is the previous command's ending point; the star ...
```
Accepting suggestion but also removing the existing ";"

dipikabh committed:
Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

dipikabh commented:
Hi @estelle, thanks for your thorough review, I appreciate it.

I've accepted some of your suggestions from the last set, but I'll have to address the remaining feedback and comments in a new PR. I've been delayed in delivering the `shape()` page as part of the Firefox release.

To allow this one to land, I'll need to ask somebody else from the content team for approval. I'm happy to work on further iterations on the page in a follow-up PR.

pepelsbey reviewed (approved):
Looks good to me. Great work!

Let’s merge this one first and then continue improving it if needed.

dipikabh commented:
@estelle, fyi - I'm working on a new PR to address your other comments. Will link to this PR soon.

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+clip-path: shape(nonzero from 0 0, line to 10px 10px);
+
+/* <move-command>, <line-command>, and close */
+offset-path: shape(from 10px 10px, move by 10px 5px, line by 20px 40%, close);
```
Yes, this is intentional...to indicate the starting position for the next command without drawing anything. This is also demo'ed in the first example https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape#using_shape_to_define_a_path. For further illustration, I am now adding `close` to `shape2` in the follow-up PR.

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    The syntax of most shape commands is an operation keyword, such as `move` or `line`, followed by the `by` or `to` keyword, and a set of coordinates.
+
+    `by`: Indicates that the `<coordinate-pair>` is relative to the command's starting point ("relative" value).
+
+    `to`: Indicates that the `<coordinate-pair>` is relative to the top-left corner of the reference box ("absolute" value).
```
The definition of `<coordinate-pair>` is included in the previous value [`from <coordiante-pair>`](https://pr33446.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape#from_coordinate-pair)

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+
+    The following `<shape-command>`s can be specified:
+
+    - `<move-command>`: Specified as `move [by | to] <coordinate-pair>`. This command adds a [MoveTo command](/en-US/docs/Web/SVG/Attribute/d#moveto_path_commands) to the list of shape commands. It draws nothing. Rather, it specifies the starting position for the next command. The `by` or `to` keyword specifies whether the `<coordinate-pair>` point is "relative" or "absolute", respectively. If th ...
```
Removed the quotes after the first definition.

To demo the `<move-command>` after the `close` command, made some changes to the first example. Updated the animation and for `shape1`, added a `close` and a `move` after this `close` (also added an `hline` after `move` to show the impact of `move`).

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+  vline by 60px
+);
+
+clip-path: shape(
```
Done in https://pr33783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape/shape#syntax

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +137,25 @@ The values in a `<basic-shape>` function are computed as specified, with these e
+
+- **Both shapes are of type `path()`**: Interpolation is applied to each parameter as a {{cssxref("&lt;number&gt;")}} if the path strings in both the shapes match the number, type, and sequence of [path data commands](/en-US/docs/Web/SVG/Attribute/d#path_commands).
+
+- **Both shapes are of type `shape()`**: Interpolation is applied between each corresponding value if they have the identical command keyword and use the same `<by-to>` keyword. If `shape()` is used in the {{cssxref("clip-path")}} property, the two shapes interpolate if they also have the same `<fill-rule>`.
```
Instead of adding this to the interpolation section, I've added this sentence to the syntax section:

dipikabh replied:
Oh, you do have that as your next suggestion

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
 ```
 
-Defines a radius for a circle or ellipse. If omitted it defaults to `closest-side`.
+The `from <coordinate-pair>` parameter represents the starting point for the first shape command, and `<shape-command>` defines one one or more shape commands, which are similar to the [SVG path commands](/en-US/docs/Web/SVG/Attribute/d#path_commands).
````
Browser compact for both `path()` https://developer.mozilla.org/en-US/docs/Web/CSS/basic-shape/path#browser_compatibility and `shape-outside` https://developer.mozilla.org/en-US/docs/Web/CSS/shape-outside#browser_compatibility indicate that `path()` is not supported in `shape-outside`. I also do not see `path()` working in `shape-outside` now that you mention it. 

I've added some notes to the issue you opened for `path()`: https://github.com/mdn/content/issues/33593#issuecomment-2135771704

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
-    ```css
-    ellipse( [ <shape-radius>{2} ]? [ at <position> ]? )
-    ```
+The `<shape-radius>` argument represents the radius of the circle defined as either a {{cssxref("length")}} or a {{cssxref("percentage")}}. A percentage value here is resolved from the used width and height of the reference box as `sqrt(width^2+height^2)/sqrt(2)`. If omitted, the radius is defined by `closest-side`.
````
The explanation for `closest-side` is included in on the same page under [Common parameters](https://developer.mozilla.org/en-US/docs/Web/CSS/basic-shape#common_parameters). Linking to a different section might change the context for the readers.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
````diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+```
 
-- `{{cssxref("path","path()")}}`
+The function takes a list of comma-separated coordinate pairs, each consisting of two space-separated `<length-percentage>` values as the _xi_ and _yi_ pair. These values represents the x and y axis coordinates of the vertex of the polygon at position _i_.
````
Updated as:

> These values represent the x and y axis coordinates of the polygon at position _i_ (the vertex point where two lines meet).

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
+When creating a shape, the reference box is defined by the property that uses `<basic-shape>` values. The coordinate system for the shape has its origin at the top-left corner of the element's margin box by default, with the x-axis running to the right and the y-axis running downwards. All the lengths expressed in percentages are resolved from the dimensions of the reference box.
 
-`farthest-side` uses the length from the center of the shape to the farthest side of the reference box. For circles, this is the farthest side in any dimension. For ellipses, this is the farthest side in the radius dimension.
+The default reference box is the [`margin-box`](/en-US/docs/Web/CSS/box-edge#margin-box), as demonstrated in the image below. The image shows a circle created using `shape-outside: circle(50%)` and inspected using a browser's Developer Tools, highlighting the different parts of the box model. The shape here is defined with reference to the margin-box.
```
Fixed in https://pr33783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape#description

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -134,14 +137,25 @@ The values in a `<basic-shape>` function are computed as specified, with these e
 ### Interpolation of basic shapes
 
-When animating between one `<basic-shape>` and another, the rules below are applied. The values in the shape functions {{Glossary("interpolation", "interpolate")}} as a simple list. The list values interpolate as {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. If list values are not one of those types but are identical, those valu ...
+When animating from one `<basic-shape>` to another, the {{Glossary("interpolation")}} rules listed below are followed. For any interpolation to happen between two shapes, both must use the same reference box. The values between two `<basic-shape>` functions interpolate based on their computed values, forming a list. The values in the list are interpolated as {{cssxref("&lt;number&gt;")}}, {{cssxr ...
```
Switched the position of 2nd and 3rd sentences for flow. Updated as:

> When animating between two `<basic-shape>` functions, the {{Glossary("interpolation")}} rules listed below are followed. The parameter values of each `<basic-shape>` function form a list. For interpolation to occur between two shapes, both shapes must use the same reference box and the number and type of values in both `<basic-shape>` lists must match.
> 
> Each value in the lists of the two `<basic-shape>` functions is interpolated based on its computed value as a {{cssxref("&lt;number&gt;")}}, {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, {{cssxref("&lt;angle&gt;")}}, or {{cssxref("calc", "calc()")}} where possible. Interpolation can still occur if the values are not one of those data types but are identical between the two interpolating basic shape functions, such as `nonzero`.

dipikabh replied:
Updated in https://github.com/mdn/content/pull/33783

dipikabh commented on files/en-us/web/css/basic-shape/shape/index.md:
```diff
@@ -0,0 +1,278 @@
+}
+
+.shape1 {
+  offset-path: shape(from 30% 60px, curve to 180px 180px via 90px 190px);
```
The use of a mix of % and length unit is intentional.

dipikabh commented on files/en-us/web/css/basic-shape/index.md:
```diff
@@ -13,116 +13,119 @@ The **`<basic-shape>`** [CSS](/en-US/docs/Web/CSS) [data type](/en-US/docs/Web/C
-## Description
+![An image showing a circle inspected with the Firefox DevTools Shape Inspector. The different parts of the box model are highlighted.](shapes-reference-box.png)
 
 ### Computed values of basic shapes
```
Please see if this rewrite helps: https://pr33783.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/basic-shape#computed_values_of_basic_shapes
